### PR TITLE
Feat add environment client

### PR DESCRIPTION
**Pull request by alonnoga**
### Issue & Steps to Reproduce / Feature Request
[//]: <> (choose “resolves” “fixes” or “closes” and put link for the issue)
adding api client for environments

I'm gonna split the ticket.
first part - basic environment CRUD, with only part of the fields on the resource 
second part - rest of the fields ( some field updates require different endpoints such as updateEnvironemntTTL, deployEnvironment etc)

### CODE REVIEW

**Comment by alonnoga** [comment]
[Terraform provider - Environment Resource](https://trello.com/c/vCEIi9gh/3988-terraform-provider-environment-resource)

File: client/model.go [hunk 1]
```diff
@@ -40,6 +40,73 @@ type ProjectCreatePayload struct {
...
+	BlueprintId          string                `json:"blueprintId,omitempty"`
+	BlueprintRevision    string                `json:"blueprintRevision,omitempty"`
+	BlueprintRepository  string                `json:"blueprintRepository,omitempty"`
+	ConfigurationChanges *ConfigurationChanges `json:"configurationChanges,omitempty"`
+	TTL                  *TTL                  `json:"ttl,omitempty"`
+	EnvName              string                `json:"envName,omitempty"`
+	UserRequiresApproval bool                  `json:"userRequiresApproval,omitempty"`
+	Targets              string                `json:"targets,omitempty"`
+	PrPlanOnly           bool                  `json:"prPlanOnly,omitempty"`
+	// TODO: not sure about this one
```
Comment (alonnoga) [inline client/model.go:65]: not sure about the type, gonna test it when I add deployments on update

File: client/model.go [hunk 2]
```diff
@@ -40,6 +40,73 @@ type ProjectCreatePayload struct {
...
+	Type        string                      `json:"type"`
+	ToDelete    string                      `json:"toDelete,omitempty"`
+	Schema      *ConfigurationChangesSchema `json:"schema,omitempty"`
+	Description string                      `json:"description,omitempty"`
+	IsSensitive string                      `json:"isSensitive,omitempty"`
+}
+
+type ConfigurationChangesSchema struct {
+	Type string `json:"type,omitempty"`
+	// TODO: not sure this works
```
Comment (alonnoga) [inline client/model.go:96]: like the one above (this is part of [deployment request](https://developer.env0.com/docs/api/b3A6MTkzNDg1MTk-create-environment))

File: client/model.go [hunk 3]
```diff
@@ -40,6 +40,73 @@ type ProjectCreatePayload struct {
 	Description string `json:"description"`
 }
 
+type EnvironmentCreate struct {
```
Comment (eranelbaz) [inline client/model.go:43]: what about the TTL here?

File: client/model.go [hunk 4]
```diff
@@ -40,6 +40,73 @@ type ProjectCreatePayload struct {
...
+	BlueprintRepository  string                `json:"blueprintRepository,omitempty"`
+	ConfigurationChanges *ConfigurationChanges `json:"configurationChanges,omitempty"`
+	TTL                  *TTL                  `json:"ttl,omitempty"`
+	EnvName              string                `json:"envName,omitempty"`
+	UserRequiresApproval bool                  `json:"userRequiresApproval,omitempty"`
+	Targets              string                `json:"targets,omitempty"`
+	PrPlanOnly           bool                  `json:"prPlanOnly,omitempty"`
+	// TODO: not sure about this one
+	CustomEnv0EnvironmentVariables *interface{} `json:"customEnv0EnvironmentVariables,omitempty"`
+	GitUserData                    *GitUserData `json:"gitUserData,omitempty"`
```
Comment (eranelbaz) [inline client/model.go:67]: Same as pr plan 
only for VCS events we can remove that

File: client/model.go [hunk 5]
```diff
@@ -40,6 +40,73 @@ type ProjectCreatePayload struct {
...
+type DeployRequest struct {
+	BlueprintId          string                `json:"blueprintId,omitempty"`
+	BlueprintRevision    string                `json:"blueprintRevision,omitempty"`
+	BlueprintRepository  string                `json:"blueprintRepository,omitempty"`
+	ConfigurationChanges *ConfigurationChanges `json:"configurationChanges,omitempty"`
+	TTL                  *TTL                  `json:"ttl,omitempty"`
+	EnvName              string                `json:"envName,omitempty"`
+	UserRequiresApproval bool                  `json:"userRequiresApproval,omitempty"`
+	Targets              string                `json:"targets,omitempty"`
+	PrPlanOnly           bool                  `json:"prPlanOnly,omitempty"`
```
Comment (eranelbaz) [inline client/model.go:64]: Pr Plan Only is only for vcs Events
isn't required so we can remove that

File: client/model.go [hunk 6]
```diff
@@ -40,6 +40,73 @@ type ProjectCreatePayload struct {
...
+	// TODO: not sure about this one
+	CustomEnv0EnvironmentVariables *interface{} `json:"customEnv0EnvironmentVariables,omitempty"`
+	GitUserData                    *GitUserData `json:"gitUserData,omitempty"`
+	TriggerName                    string       `json:"triggerName,omitempty"`
```
Comment (eranelbaz) [inline client/model.go:68]: not for this pr but this one should be `User`

File: client/model.go (hunk 1 above)
Comment (liranfarage89) [reply client/model.go:65]: maybe `map[string]interface{}` ?
https://stackoverflow.com/questions/40559250/golang-dynamically-creating-member-of-struct

File: client/model.go [hunk 7]
```diff
@@ -40,6 +40,73 @@ type ProjectCreatePayload struct {
...
+	PullRequestPlanDeployments  bool           `json:"pullRequestPlanDeployments,omitempty"`
+	AutoDeployOnPathChangesOnly bool           `json:"autoDeployOnPathChangesOnly,omitempty"`
+	AutoDeployByCustomGlob      bool           `json:"autoDeployByCustomGlob,omitempty"`
+}
+
+type DeployRequest struct {
+	BlueprintId          string                `json:"blueprintId,omitempty"`
+	BlueprintRevision    string                `json:"blueprintRevision,omitempty"`
+	BlueprintRepository  string                `json:"blueprintRepository,omitempty"`
+	ConfigurationChanges *ConfigurationChanges `json:"configurationChanges,omitempty"`
```
Comment (eranelbaz) [inline client/model.go:59]: this one should be array of them

File: client/model.go (hunk 2 above)
Comment (liranfarage89) [reply client/model.go:96]: this one looks like the same object in `api.d.ts` so I think it's ok and should work.

`type PartialJSONSchema7 = { type?: 'string'; enum?: string[] };`

File: client/environment_test.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+					})
+
+				environment, err = apiClient.Environment(mockEnvironment.Id)
+			})
+
+			It("Should send GET request", func() {
+				httpCall.Times(1)
+			})
+
+			It("Should return teams", func() {
```
Comment (eranelbaz) [inline client/environment_test.go:79]:
```suggestion
			It("Should return environment", func() {
```

**Review by eranelbaz — COMMENTED** [review]
Looking good
just lots of question regrading the types

File: client/model.go (hunk 3 above)
Comment (alonnoga) [reply client/model.go:43]: gonna be added [here](https://trello.com/c/K68z3UXy/4784-terraform-provider-environment-resource-part-2)

File: client/model.go (hunk 6 above)
Comment (alonnoga) [reply client/model.go:68]: you mean on the resource?

File: client/model.go (hunk 7 above)
Comment (alonnoga) [reply client/model.go:59]: so turns out it can be both, so i'll make it an array so were covered

File: client/environment_test.go (hunk 8 above)
Comment (alonnoga) [reply client/environment_test.go:79]: thanks

File: client/model.go (hunk 7 above)
Comment (alonnoga) [reply client/model.go]: 63d1a3c

File: client/model.go (hunk 5 above)
Comment (alonnoga) [reply client/model.go]: 63d1a3c

File: client/model.go (hunk 4 above)
Comment (alonnoga) [reply client/model.go]: 63d1a3c

File: client/model.go (hunk 1 above)
Comment (alonnoga) [reply client/model.go:64]: I will change it on the pr for [this ticket](https://trello.com/c/K68z3UXy/4784-terraform-provider-environment-resource-part-2) that way I'll also actually test it

File: client/model.go (hunk 6 above)
Comment (eranelbaz) [reply client/model.go:66]: the value of `TriggerName` should be `User`

**Review by eranelbaz — APPROVED** [review]

**Pull request merged**